satellite/gracefulexit/validation.go

```diff
@@ -0,0 +1,116 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package gracefulexit
+
+import (
+	"bytes"
+	"context"
+
+	"storj.io/storj/pkg/identity"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/signing"
+	"storj.io/storj/uplink/eestream"
+)
+
+func validatePointer(ctx context.Context, pointer *pb.Pointer, incomplete *TransferQueueItem) (*pb.RemotePiece, error) {
+	remote := pointer.GetRemote()
+	nodeID := incomplete.NodeID
+
+	pieces := remote.GetRemotePieces()
+	var nodePiece *pb.RemotePiece
+	for _, piece := range pieces {
+		if piece.NodeId == nodeID && piece.PieceNum == incomplete.PieceNum {
+			nodePiece = piece
+		}
+	}
+
+	if nodePiece == nil {
+		return nil, Error.New("piece no longer held by node")
+	}
+
+	return nodePiece, nil
+}
+
+func validateRedundancyThreshold(ctx context.Context, pointer *pb.Pointer, redundancy eestream.RedundancyStrategy) bool {
+	pieces := pointer.GetRemote().GetRemotePieces()
+	if len(pieces) > redundancy.OptimalThreshold() {
+		return false
+	}
+
+	return true
+}
+
+func validatePendingTransfer(ctx context.Context, transfer *pendingTransfer) error {
+	if transfer.satelliteMessage == nil {
+		return Error.New("Satellite message cannot be nil")
+	}
+	if transfer.satelliteMessage.GetTransferPiece() == nil {
+		return Error.New("Satellite message transfer piece cannot be nil")
+	}
+	if transfer.satelliteMessage.GetTransferPiece().GetAddressedOrderLimit() == nil {
+		return Error.New("Addressed order limit on transfer piece cannot be nil")
+	}
+	if transfer.satelliteMessage.GetTransferPiece().GetAddressedOrderLimit().GetLimit() == nil {
+		return Error.New("Addressed order limit on transfer piece cannot be nil")
+	}
+	if transfer.path == nil {
+		return Error.New("Transfer path cannot be nil")
+	}
+	if transfer.originalPointer == nil || transfer.originalPointer.GetRemote() == nil {
+		return Error.New("could not get remote pointer from transfer item")
+	}
+
+	return nil
+}
+
+func verifyPieceTransferred(ctx context.Context, message *pb.StorageNodeMessage_Succeeded, transfer *pendingTransfer, satellite signing.Signer, nodeID *identity.PeerIdentity) error {
```

Comment:
nit - change `nodeID *identity.PeerIdentity` to `receivingID *identity.PeerIdentity`